src/dbnode/storage/index.go

```diff
@@ -343,28 +446,33 @@ func (i *nsIndex) WriteBatch(
 
 // WriteBatches is called by the indexInsertQueue.
 func (i *nsIndex) writeBatches(
-	batches []*index.WriteBatch,
+	batch *index.WriteBatch,
 ) {
 	// NB(prateek): we use a read lock to guard against mutation of the
 	// indexBlocks, mutations within the underlying blocks are guarded
 	// by primitives internal to it.
 	i.state.RLock()
-	defer i.state.RUnlock()
 	if !i.isOpenWithRLock() {
+		i.state.RUnlock()
 		// NB(prateek): deliberately skip calling any of the `OnIndexFinalize` methods
 		// on the provided inserts to terminate quicker during shutdown.
 		return
 	}
-
 	now := i.nowFn()
 	futureLimit := now.Add(1 * i.bufferFuture)
 	pastLimit := now.Add(-1 * i.bufferPast)
-	writeBatchFn := i.writeBatchForBlockStartWithRLock
-	for _, batch := range batches {
-		// Ensure timestamp is not too old/new based on retention policies and that
-		// doc is valid.
-		batch.ForEach(func(idx int, entry index.WriteBatchEntry,
+	// NB(r): Release lock early to avoid writing batches impacting ticking
+	// speed, etc.
+	// Sometimes foreground compaction can take a long time during heavy inserts.
+	// Each lookup to ensureBlockPresent checks that index is still open, etc.
```

Comment:
This comment could use a little clarification too. Looks like notes for a comment to write later lol